File: packages/guards/src/pii-ar.ts
import type {
  Guard,
  GuardResult,
  GuardContext,
} from 'open-guardrail-core';

type PiiArEntity =
  | 'national-id'
  | 'passport'
  | 'phone'
  | 'iban';

interface PiiArOptions {
  entities: PiiArEntity[];
  action: 'block' | 'warn' | 'mask';
}

interface PiiArMatch {
  type: PiiArEntity;
  value: string;
  start: number;
  end: number;
}

const PATTERNS: Record<PiiArEntity, RegExp> = {
  'national-id': /\b\d{10}\b/g,
  passport: /\b[A-Z]\d{8}\b/g,
  phone: /\b(?:\+?(?:966|971|20|962|961|212|216)[-.\s]?)?\d{1,3}[-.\s]?\d{3,4}[-.\s]?\d{4}\b/g,
  iban: /\b[A-Z]{2}\d{2}[A-Z0-9]{4}\d{7}([A-Z0-9]?){0,16}\b/g,
};

const MASK_LABELS: Record<PiiArEntity, string> = {
  'national-id': '[رقم الهوية]',
  passport: '[رقم جواز السفر]',
  phone: '[رقم الهاتف]',
  iban: '[رقم الحساب البنكي]',
};

function detect(text: string, entities: PiiArEntity[]): PiiArMatch[] {
  const matches: PiiArMatch[] = [];
  for (const entity of entities) {
    const re = new RegExp(PATTERNS[entity].source, 'g');
    let m: RegExpExecArray | null;
    while ((m = re.exec(text)) !== null) {
      matches.push({ type: entity, value: m[0], start: m.index, end: m.index + m[0].length });
    }
  }
  return matches.sort((a, b) => b.start - a.start);
}

function maskText(text: string, matches: PiiArMatch[]): string {
  let result = text;
  for (const m of matches) {
    result = result.slice(0, m.start) + MASK_LABELS[m.type] + result.slice(m.end);
  }
  return result;
}

export function piiAr(options: PiiArOptions): Guard {
  return {
    name: 'pii-ar',
    version: '0.1.0',
    description: 'Arabic/Middle East PII detection and masking',
    category: 'privacy',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const matches = detect(text, options.entities);
      const triggered = matches.length > 0;
      if (!triggered) return { guardName: 'pii-ar', passed: true, action: 'allow', latencyMs: Math.round(performance.now() - start) };
      if (options.action === 'mask') {
        return { guardName: 'pii-ar', passed: true, action: 'override', overrideText: maskText(text, matches), latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type, value }) => ({ type, value })) } };
      }
      return { guardName: 'pii-ar', passed: false, action: options.action, latencyMs: Math.round(performance.now() - start), details: { detected: matches.map(({ type, value }) => ({ type, value })) } };
    },
  };
}

File: python/open_guardrail/guards/pii_ar.py
"""Arabic/Middle East PII detection and masking guard."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS: dict[str, re.Pattern[str]] = {
    "national-id": re.compile(r"\b[12]\d{9}\b"),
    "iqama": re.compile(r"\b2\d{9}\b"),
    "passport-ar": re.compile(r"\b[A-Z]\d{8}\b"),
    "phone-ar": re.compile(r"\b(?:\+(?:966|971|20|962|961|974|973|968|965)|00(?:966|971|20))\s?\d[\s\-]?\d{3}[\s\-]?\d{4}\b"),
    "email": re.compile(r"\b[A-Za-z0-9._%+\-]+@[A-Za-z0-9.\-]+\.[A-Za-z]{2,}\b"),
}

_MASK_LABELS: dict[str, str] = {
    "national-id": "[رقم الهوية]",
    "iqama": "[رقم الإقامة]",
    "passport-ar": "[جواز السفر]",
    "phone-ar": "[رقم الهاتف]",
    "email": "[البريد الإلكتروني]",
}


def _detect(text: str, entities: list[str]) -> list[dict]:
    matches: list[dict] = []
    for entity in entities:
        pat = _PATTERNS.get(entity)
        if not pat:
            continue
        for m in pat.finditer(text):
            matches.append({"type": entity, "value": m.group(), "start": m.start(), "end": m.end()})
    matches.sort(key=lambda x: x["start"], reverse=True)
    return matches


def _mask(text: str, matches: list[dict]) -> str:
    result = text
    for m in matches:
        result = result[: m["start"]] + _MASK_LABELS.get(m["type"], "[PII]") + result[m["end"] :]
    return result


class _PiiAr:
    def __init__(self, *, action: str = "block", entities: Optional[List[str]] = None) -> None:
        self.name = "pii-ar"
        self.action = action
        self._entities = entities or list(_PATTERNS.keys())

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matches = _detect(text, self._entities)
        elapsed = (time.perf_counter() - start) * 1000
        if not matches:
            return GuardResult(guard_name="pii-ar", passed=True, action="allow", latency_ms=round(elapsed, 2))
        detected = [{"type": m["type"], "value": m["value"]} for m in matches]
        if self.action == "mask":
            return GuardResult(guard_name="pii-ar", passed=True, action="override", override_text=_mask(text, matches), latency_ms=round(elapsed, 2), details={"detected": detected})
        return GuardResult(guard_name="pii-ar", passed=False, action=self.action, message="Arabic PII detected", latency_ms=round(elapsed, 2), details={"detected": detected})


def pii_ar(*, action: str = "block", entities: Optional[List[str]] = None) -> _PiiAr:
    return _PiiAr(action=action, entities=entities)
